tachyon/math/matrix/matrix_utils.h

```diff
@@ -199,6 +200,42 @@ std::vector<ExtField> ColumnwiseDotProduct(
   return ret;
 }
 
+template <typename ExtField, typename Derived>
+std::vector<ExtField> DotExtPowers(const Eigen::MatrixBase<Derived>& mat,
+                                   const ExtField& base) {
+  using F = typename ExtensionFieldTraits<ExtField>::BaseField;
+  using PackedField = typename PackedFieldTraits<F>::PackedField;
+  using ExtendedPackedField =
+      typename ExtendedPackedFieldTraits<ExtField>::ExtendedPackedField;
+  size_t rows = static_cast<size_t>(mat.rows());
+  size_t degree = ExtField::ExtensionDegree();
+  size_t packed_n = PackedField::N;
+  std::vector<ExtendedPackedField> packed_ext_powers =
+      ExtField::GetExtendedPackedPowers(
+          base, ((mat.cols() + packed_n - 1) / packed_n) * packed_n);
+  std::vector<ExtField> ret(rows);
+
+  OPENMP_PARALLEL_FOR(size_t r = 0; r < rows; ++r) {
+    std::vector<PackedField> row_packed = PackRowHorizontallyPadded(mat.row(r));
+    ExtendedPackedField packed_sum_of_packed(ExtendedPackedField::Zero());
+    for (size_t i = 0; i < row_packed.size(); ++i) {
+      packed_sum_of_packed +=
+          packed_ext_powers[i] * ExtendedPackedField(row_packed[i]);
+    }
```

Comment:
I don't seem to quite understand. Could you explain which parts are being converted repetitively?

Reply:
First, each row is converted to `std::vector<PackedField>` and then basically to `std::vector<ExtendedPackedField>`

Reply:
Ah I see; however, since ExtendedPackedField only take in a packed field type, and all the general data is in field form, the conversion from field/num to packedfield to ExtendedPackedField is inevitable. I don't think any conversions are being repeated though.

Reply:
Hmm I see! I will look a bit more and resolve :)